linera-core/src/client.rs

```diff
@@ -364,13 +365,73 @@ where
     }
 
     /// Listens to notifications about the current chain from all validators.
-    pub async fn listen(&mut self) -> Result<NotificationStream> {
-        let committee = self.local_committee().await?;
+    pub async fn listen(this: Arc<Mutex<Self>>) -> Result<NotificationStream>
+    where
+        P: Send + 'static,
+    {
+        let chain_id = this.lock().await.chain_id;
+        let committee = this.lock().await.local_committee().await?;
+        let nodes = this
+            .lock()
+            .await
+            .validator_node_provider
+            .make_nodes(&committee)?;
         let mut streams = Vec::new();
-        for (name, mut node) in self.validator_node_provider.make_nodes(&committee)? {
-            match node.subscribe(vec![self.chain_id]).await {
-                Ok(notification_stream) => {
-                    streams.push(notification_stream);
+        for (name, mut node) in nodes {
+            let this = this.clone();
+            match node.subscribe(vec![chain_id]).await {
+                Ok(stream) => {
+                    let stream = stream.filter(move |notification| {
+                        let notification = notification.clone();
+                        let this = this.clone();
+                        let node = node.clone();
+                        Box::pin(async move {
+                            // TODO(#687): skip notifications that are not useful.
+                            let mut client = this.lock().await;
+                            match &notification.reason {
+                                Reason::NewMessage { origin, height } => {
+                                    if let Err(e) = client.find_received_certificates_from_validator(name, node).await {
```

Comment:
Should `find_received_certificates_from_validator` also take `Arc<Mutex<Self>>` instead, so we don't need to keep the lock while waiting for responses?